### ISSUE TITLE
Strange HttpClient behavior with many concurrent requests on OS X and Linux

### ISSUE THREAD

**Issue by joelverhagen** [dotnet/corefx#7561]
When a lot of concurrent requests are thrown at `HttpClient`, I see varying behavior on the different platforms:

**Mac OS X**: lots of "Couldn't connect to server" and "Couldn't resolve host name" errors.
**Linux**: works a lot of the time but sometimes hangs due and throws a `TaskCancelledException` due to `HttpClient.Timeout`.
**Windows**: works fine.

Here is a small `dotnet` application that reproduces the problem: [lotsofhttp.zip](https://github.com/dotnet/corefx/files/207394/lotsofhttp.zip)

The input I give to my program is:
```
dotnet run 500 10 500 https://api.nuget.org/v3-flatcontainer/newtonsoft.json/index.json
```

That is, 500 concurrent tasks, 10 iterations per tasks, and a throttling semaphore of 500. The URL `https://api.nuget.org/v3-flatcontainer/newtonsoft.json/index.json` is downloaded for each iteration.

My suspicion is there is some kind of handle limiting going on OS X that prevents so many concurrent requests.

Are these failures expected? Am I voiding a warranty :smile:? The workaround is certainly to apply throttling above `HttpClient` in the application, correct?

I am using System.Net.Http 4.0.1-rc2-23931.

**Comment by stephentoub** [comment]
> My suspicion is there is some kind of handle limiting going on OS X that prevents so many concurrent requests.

Do you get these failures if you raise the max open file limit before running dotnet run, e.g.
```
ulimit -n 8192
```
?

**Comment by joelverhagen** [comment]
Yes, that fixes the problem. In fact, `ulimit -n 600` fixes the input provided above.

**Comment by stephentoub** [comment]
> Yes, that fixes the problem.

Thanks for confirming.  This is an underlying OS limitation, and I'm not sure how much we can realistically do about it.  Some kind of throttling would need to be done, and/or ensuring that the limit is always set high enough.

One thing we could consider would be throttling the max number of connections libcurl is allowed to use, each one of which will have a socket, which is an open file.  We could query for the allowed max number of open files, and backoff from that by some guessed amount, then set that as an upper bound on the number of connections allowed.  I'm not sure how good an idea that is, though.  

Thoughts?  cc'ing some folks who might have an opinion:
@ericeil, @kapilash, @bartonjs, @ellismg

**Comment by stephentoub** [comment]
(Also, @joelverhagen, FYI, I believe there are some stress bugs in some older versions of libcurl related to DNS resolution.  When I run your test on an Ubuntu install with curl v7.35, I got ~99% success rate, with the remaining requests failing due to the cited error messages; when I use v7.49, I get 100% success.)

**Comment by ericeil** [comment]
Even if we implemented some sort of heuristic, as @stephentoub suggests, it would only be effective for cases that *only* use HttpClient.  If, for example, you were also streaming the response data to a local file, that would double the number of file descriptors involved.  If we tried to compensate for *that*, then we'd find another scenario that saved to a local file *and* forwarded the data to a raw TCP socket... etc., etc.  There's no way in general that we can anticipate how many file descriptors might be in use *around* any single operation, so I don't see how we could throttle in any generally meaningful way.

(Even if we had throttling built into the PAL for *all* file descriptors, it wouldn't cover files unknown to CoreFx, such as cases of user code doing direct p/invokes.)

I think the right thing is to simply make sure that CoreFx/CoreCLR only use a "reasonable," predictable, and easily justifiable number of file descriptors for any given operation, as seems to be the case here, and then let users decide whether they need to throttle operations or increase the limits, depending on their specific scenario.
